compaction/compaction_manager.cc

```diff
@@ -1530,19 +1531,24 @@ future<> compaction_manager::perform_cleanup(owned_ranges_ptr sorted_owned_range
             t.schema()->ks_name(), t.schema()->cf_name()));
     }
 
-    if (sorted_owned_ranges->empty()) {
-        throw std::runtime_error("cleanup request failed: sorted_owned_ranges is empty");
-    }
-
+    // Called with compaction_disabled
     auto get_sstables = [this, &t, sorted_owned_ranges] () -> future<std::vector<sstables::shared_sstable>> {
         return seastar::async([this, &t, sorted_owned_ranges = std::move(sorted_owned_ranges)] {
-            auto sstables = std::vector<sstables::shared_sstable>{};
-            const auto candidates = get_candidates(t);
-            std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(sstables), [&sorted_owned_ranges] (const sstables::shared_sstable& sst) {
-                seastar::thread::maybe_yield();
-                return update_sstable_cleanup_state(sst, *sorted_owned_ranges);
-            });
-            return sstables;
+            auto& cs = get_compaction_state(&t);
+            auto filter_candidates = [&cs, &sorted_owned_ranges] (const lw_shared_ptr<sstables::sstable_list>& candidates) {
+                for (const auto& sst : *candidates) {
+                    if (update_sstable_cleanup_state(sst, *sorted_owned_ranges)) {
+                        cs.sstables_requiring_cleanup.insert(sst);
```

Comment:
It seems to me that we don't need to store the cleanup state in sstables themselves.

If we make compaction manager the authorative source, then we make it more robust to avoid divergences on if a sstable actually needs cleanup. 

A compaction will only enable filtering reader if owned_ranges is provided and any of the input is in compaction_state::sstables_requiring_cleanup.

Any outside users, like reshard on refresh, can use a compaction_manager api, sstable_requires_cleanup(), which consults compaction_state::sstables_requiring_cleanup.
update_sstable_cleanup_state() would go into compaction manager and implementation becomes inserting sstable into above list if cleanup is needed. This way, we avoid divergence and have a single place to go for checking the state.
Another benefit is that sstable doesn't have to know if cleanup is needed, which is a good move to not mix distributed with storage layer.

Reply:
I'd like to use the requires_cleanup state also to make a filtering reader for normal sstable readers, not just for compaction, as we need to skip the disowned tokens everywhere - like in range reads (assuming single partition reads will typically not access disowned tokens, and if they do its since they race with topology changes, and we can't do anything about it)

Reply:
What about:

```
bool
replica::table::requires_cleanup(const sstables::sstable_set& set) const {
    return bool(set.for_each_sstable_until([this] (const shared_sstable& sst) {
        auto& cg = compaction_group_for_sstable(sst);
        // compaction_manager::requires_cleanup() only looks up sstable in compaction_state, which is average O(1) (std::unordered_set<...>::count()).
        return stop_iteration(_compaction_manager.requires_cleanup(cg.as_table_state()));
    }));
}
```

Then we can kill sstable_set::needs_cleanup() which is good because sstable_set doesn't need to know about cleanup.

Reply:
This will work.  That's yet another example why compaction group is a bad name IMO and I'd really prefer sub_table or something like that.
Since the requires_cleanup state essentially says that the sstable might hold disowned tokens. They need filtering for both regular reads and for compaction, but the root issue here is the need for filtering, not about cleanup compaction.